explicit tail calls: hang in const eval of recursive call

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
#![expect(incomplete_features)]
#![feature(explicit_tail_calls)]

pub const fn f() {
    become g();
}

const fn g() {
    become g();
}

const _: () = f();

fn main() {}
```

It seems there is no output here until I ctrl+c this.
`perf top` shows some activity in `_RNvMs_NtNtCs69nEItziXW1_16rustc_const_eval9interpret4callINtNtB6_12eval_context8InterpCxNtNtNtB8_10const_eval7machine18CompileTimeMachineE16init_stac`

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.90.0-nightly (adcb3d3b4 2025-07-31)
binary: rustc
commit-hash: adcb3d3b4cd3b7c4cde642f3ed537037f293738e
commit-date: 2025-07-31
host: x86_64-unknown-linux-gnu
release: 1.90.0-nightly
LLVM version: 20.1.8
```

```
^Cerror[E0080]: compilation was interrupted
  --> a.rs:12:15
   |
12 | const _: () = f();
   |               ^^^ evaluation of `_` failed inside this call
   |
note: inside `g`
  --> a.rs:9:5
   |
 9 |     become g();
   |     ^^^^^^^^^^ the failure occurred here

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0080`.
```

--- a/compiler/rustc_mir_transform/src/ctfe_limit.rs
+++ b/compiler/rustc_mir_transform/src/ctfe_limit.rs
@@ -18,7 +18,7 @@ impl<'tcx> crate::MirPass<'tcx> for CtfeLimit {
             .basic_blocks
             .iter_enumerated()
             .filter_map(|(node, node_data)| {
-                if matches!(node_data.terminator().kind, TerminatorKind::Call { .. })
+                if matches!(node_data.terminator().kind, TerminatorKind::Call { .. } | TerminatorKind::TailCall { .. })
                     // Back edges in a CFG indicate loops
                     || has_back_edge(doms, node, node_data)
                 {

--- a/tests/ui/explicit-tail-calls/infinite-recursion-in-ctfe.rs
+++ b/tests/ui/explicit-tail-calls/infinite-recursion-in-ctfe.rs
@@ -1,0 +1,10 @@
+#![feature(explicit_tail_calls)]
+#![expect(incomplete_features)]
+
+const _: () = f();
+
+const fn f() {
+    become f(); //~ error: constant evaluation is taking a long time
+}
+
+fn main() {}

--- a/tests/ui/explicit-tail-calls/infinite-recursion-in-ctfe.stderr
+++ b/tests/ui/explicit-tail-calls/infinite-recursion-in-ctfe.stderr
@@ -1,0 +1,17 @@
+error: constant evaluation is taking a long time
+  --> $DIR/infinite-recursion-in-ctfe.rs:7:5
+   |
+LL |     become f();
+   |     ^^^^^^^^^^
+   |
+   = note: this lint makes sure the compiler doesn't get stuck due to infinite loops in const eval.
+           If your compilation actually takes a long time, you can safely allow the lint.
+help: the constant being evaluated
+  --> $DIR/infinite-recursion-in-ctfe.rs:4:1
+   |
+LL | const _: () = f();
+   | ^^^^^^^^^^^
+   = note: `#[deny(long_running_const_eval)]` on by default
+
+error: aborting due to 1 previous error
+

